### PR TITLE
SDK Version 2.0 including new Google API features (ExposureWindows)

### PR DESCRIPTION
**Pull request by simonroesch**
- updated to play-services-nearby-exposurenotification-1.7.1-eap.aar
- use exposureWindows to compute attenuationDurations
- exposureDays now returns all exposure days, not only the last one as in previous versions
- updated defaults for attenuationBucketThresholds, new values are 55 and 63
- add Experiment-Mode to calibration app to simplify experiments with multiple devices and the new ExposureWindows-API (see CALIBRATION_APP_USAGE.md for details)
- expose EN-Module version, this can be used to extend the userAgent to be able to handle potential bugs in future EN versions from the backend
- Version 2.0.0 of the SDK will require EN module version >= 1.6, if run on older versions a notification will be generated asking the user to update Google Play Services
- add config option DP3T.setNumberOfDaysToConsiderForExposure() to define how many days after the exposure an exposure should be considered
- add config option DP3T.setNumberOfDaysToKeepExposedDays() to define how many days after an exposure is reported, this should be kept

### CODE REVIEW

File: dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/internal/AppConfigManager.java [hunk 1]
```diff
@@ -178,28 +162,41 @@ public void setAttenuationFactorMedium(float factor) {
...
+		return sharedPrefs.getLong(PREF_LAST_SYNC_CALL_TIME, 0);
 	}
 
-	public void setLastSuccessfulSyncTimes(HashMap<DayDate, Long> lastSuccessfulSyncTimes) {
-		sharedPrefs.edit().putString(PREF_LAST_SUCCESSFUL_SYNC_TIMES, Json.toJson(convertFromDateMap(lastSuccessfulSyncTimes))).apply();
+	public void setLastSyncCallTime(long time) {
+		sharedPrefs.edit().putLong(PREF_LAST_SYNC_CALL_TIME, time).apply();
+	}
+
+	public Long getLastKeyBundleTag() {
```
Comment (UBaggeler) [inline dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/internal/AppConfigManager.java:189]: I'd prefer switching to a string for the key bundle tag to have all options open in the backend to change this to something else.

Comment (simonroesch) [reply dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/internal/AppConfigManager.java]: Agreed! Has been changed to use a String value!

File: dp3t-sdk/sdk/src/androidTest/java/org/dpppt/android/sdk/internal/ExposureDayStorageTest.java [hunk 2]
```diff
@@ -87,10 +103,28 @@ public void testKeepTestsFor14DaysAfterReport() {
...
+
+	@Test
+	public void testKeepTestsFor10DaysAfterReport() {
+		ExposureDayStorage eds = ExposureDayStorage.getInstance(context);
+		eds.clear();
+		DP3T.setNumberOfDaysToKeepExposedDays(context, 10);
+		//should be returned in getExposureDays()
+		eds.addExposureDays(context,
+				Arrays.asList(new ExposureDay(-1, new DayDate().subtractDays(11), System.currentTimeMillis() - 10)));
 		//should not be considered because the report date is more than 14 days in the past
```
Comment (gallmann-ubique) [inline dp3t-sdk/sdk/src/androidTest/java/org/dpppt/android/sdk/internal/ExposureDayStorageTest.java:124]: should be "...more than 10 days..."

**Review by gallmann-ubique — APPROVED** [review]

File: .github/workflows/appcenter.yml [hunk 3]
```diff
@@ -2,7 +2,7 @@ name: Android Build
 
 on:
   push:
-    branches: [ master, master-alpha ]
+    branches: [ master, master-alpha, feature/include-google-en-api-1-5 ]
```
Comment (buenzli-ubique) [inline .github/workflows/appcenter.yml:5]: When do you plan to remove this again?

File: dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/internal/AppConfigManager.java [hunk 4]
```diff
@@ -41,38 +34,33 @@ public static synchronized AppConfigManager getInstance(Context context) {
...
 	private static final String PREF_I_AM_INFECTED_IS_RESETTABLE = "IAmInfectedIsResettable";
-	private static final String PREF_CALIBRATION_TEST_DEVICE_NAME = "calibrationTestDeviceName";
-	private static final String PREF_LAST_LOADED_TIMES = "lastLoadedTimes";
-	private static final String PREF_LAST_SYNC_CALL_TIMES = "lastExposureClientCalls";
-	private static final String PREF_LAST_SUCCESSFUL_SYNC_TIMES = "lastSuccessfulSyncTimes";
+	private static final String PREF_LAST_SYNC_CALL_TIME = "lastSyncCallTime";
+	private static final String PREF_LAST_KEY_BUNDLE_TAG = "lastKeyBundleTag";
 	private static final String PREF_DEV_HISTORY = "devHistory";
+	private static final String PREF_EN_MODULE_VERSION = "enModuleVersion";
+	private static final String PREF_NUMBERR_OF_DAYS_TO_CONSIDER_FOR_EXPOSURE = "numberOfDaysToConsiderrForExposure";
```
Comment (buenzli-ubique) [inline dp3t-sdk/sdk/src/main/java/org/dpppt/android/sdk/internal/AppConfigManager.java:41]: Typo in "numberOfDaysToConside***rr***ForExposure"

File: .github/workflows/appcenter.yml (hunk 3 above)
Comment (simonroesch) [reply .github/workflows/appcenter.yml:5]: good point, should be remove last minute before merging ;-)

**Review by buenzli-ubique — APPROVED** [review]

File: CALIBRATION_APP_USAGE.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+## Run experiment
+To start an experiment open the app and press the red "START TRACKING" button. An additional popup will show up where you have to agree to enable Exposure Notifications.
+
+At the end of the experiment go to the parameters tab (second icon at the bottom) and press "UPLOAD KEYS FOR EXPERIMENT". Make sure that the parameters from the setup phase are set correctly.
+
+To stop tracing and prevent any side-effects go back to the controls tab (first icon at the bottom) and press "STOP TRACKING".
+
+## Execute matching
+
+Make sure tracing is enabled on the device for the matching to work (this is a restriction of the Google ExposureNotification framework). To do this, go to the controls tab (first icon at the bottom) of the app and press "START TRACING" if not already active.
```
Comment (UBaggeler) [inline CALIBRATION_APP_USAGE.md:48]: "START TRACING" or "START TRACKING"? In my opinion it makes more sense to call it tracing instead of tracking.

**Review by UBaggeler — APPROVED** [review]

**Pull request merged**